bookstore/archive.py

```diff
@@ -29,6 +41,46 @@ def __init__(self, *args, **kwargs):
             s3_additional_kwargs={},
         )
 
+        self._thread_pool = ThreadPoolExecutor(max_workers=30)
+
+        self.path_locks: Dict[str, Lock] = {}
+
+    async def archive(self, record: ArchiveRecord):
+        """Process a storage write, only allowing one write at a time to each path
+        """
+        lock = self.path_locks.get(record.filepath)
+
+        try:
+            if lock is None:
+                lock = Lock()
+                self.path_locks[record.filepath] = lock
+
+            # Try to acquire the lock, and if it fails don't bother archiving
+            # as we should finish a previous write
+            await lock.acquire(1)
+
+        except TimeoutError as e:
+            self.log.info("Another archival in progress for %s", record.filepath)
+            return
+
+        # At this point we have the lock and can write to S3
+        try:
+            self.log.info("Processing storage write of %s", record.filepath)
+            full_path = self.s3_path(record.filepath)
+            await self.write_to_s3(full_path, record.content)
```

Comment:
Although, thinking about this more, since we have an `await` in here, won't this block the `archive()` call as if we just talked to S3 here directly? I was thinking we'd do a `queue.put()` in here with a low timeout, and have a separate routine that pulls from the queue and does the IO to S3.

Reply:
Tornado is on a single threaded event loop so this will allow any other archive calls, notebook server ops, etc. to go ahead. The `write_to_s3` is wrapped in a decorator that makes it return a Future that will resolve when the blocking code we've moved to threads is finished up.

Reply:
I don't follow.

Reply:
I'm still new to Tornado and Python's coroutine model but it looks like the `write_to_s3()` activity will occur in another thread but this thread that calls that will block here because of the `await` so essentially this is like a synchronous call to `write_to_s3()` here.

Reply:
It will block in the sense that code won't run below, that's true. It frees up other CPU bound work to run though, new API calls, serving assets, as well as any kernel communication processing.

The code on this branch is a little bit different, but I'll attempt to describe it.

`archive` submits the `write_to_s3` call effectively like this:

```python
future = executor.submit(write_to_s3, full_path, record.content)
await future # our own execution stops, but doesn't block other work on the ioloop
```

Reply:
Got it - thanks for the explanation - I think I've got my head around this model now :-)